File: Examples/CallingCSharpFunctionFromLua2/TestModule.cs
using System;
using LuaNET;
using LuaNET.Interop;
using LuaNET.Modules;

namespace LuaNETExample
{
    public class TestModule : LuaModule
    {
        public override void Initialize(LuaState L)
        {
            //Registers any method in this class with the LuaExternalMethod attribute so Lua can request a function pointer to C-Sharp methods
            RegisterExternalMethods();

            //This assumes testmodule.lua is in the same directory as the executable
            LuaModuleLoader.RegisterFromFile(L, "testmodule", "testmodule.lua");
        }

        [LuaExternalMethod]
        private static unsafe void TestModuleWriteLine(byte* message)
        {
            //On the lua side we define this method as taking a char pointer
            //Note that a 'char' in C-Sharp is a 16 bit entity and not 8 bit like in C/C++
            //For this reason we use a byte pointer instead
            //Lua owns this pointer (memory) and you should not free it
            string s = new string((sbyte*)message);
            Console.WriteLine(s);
        }
    }
}

File: Examples/CallingCSharpFunctionFromLua2/TestModule.lua
local ffi = require("ffi")
local luanet = require("luanet")

local library = {}
--Gets a function pointer to the registered C-Sharp function.
library.TestModuleWriteLine = luanet.findMethod("TestModuleWriteLine", "void (__cdecl*)(char*)")

local testmodule = {}

function testmodule.writeLine(message)
    if type(message) == "number" then
        message = tostring(message)
    end

    -- Converts the lua string into a C-string
    -- Lua owns this memory and will garbage collect it eventually
    local c_message = ffi.new("char[?]", #message + 1)
    ffi.copy(c_message, message)

    -- Calls the C-Sharp method
    library.TestModuleWriteLine(c_message)
end

return testmodule